pkg/cmd/pulumi/templates/org.go

```diff
@@ -0,0 +1,221 @@
+		if apiError := new(apitype.ErrorResponse); errors.As(err, &apiError) {
+			// This is what happens when we try to access org templates for an org that hasn't enabled org templates.
+			if apiError.Code == 402 {
+				logging.Infof("%q does not have access to org templates (code=%d)", org, apiError.Code)
+				return
+			}
+		} else if err != nil {
+			s.addError(err)
+			logging.Warningf("Failed to get templates from %q: %s", org, err.Error())
+			return
+		} else if orgTemplates.HasAccessError {
+			logging.Warningf("Failed to get templates from %q: Access Denied", org)
+			return
+		} else if orgTemplates.HasUpstreamError {
+			logging.Warningf("Failed to get templates from %q: Upstream Error", org)
+			return
+		}
+
+		for source, sourceTemplates := range orgTemplates.Templates {
+			logging.Infof("sourcing templates from %q", source)
+			for _, template := range sourceTemplates {
+				// These template are maintained using https://github.com/pulumi/templates, and are
+				// ingested without going through the Pulumi Cloud.
+				//
+				//
+				if strings.HasPrefix(template.TemplateURL, "https://github.com/pulumi/templates") {
+					continue
+				}
+
+				// Check if we already have this template from another source.
+				if _, ok := alreadySeenSourceURLs[template.TemplateURL]; ok {
+					// Skip a template that we have already seen.
+					continue
+				}
+				alreadySeenSourceURLs[template.TemplateURL] = struct{}{}
+
+				// If we are searching for a template of a specific name,
+				// only match templates of that name.
+				if templateName != "" && templateName != template.Name {
+					logging.V(10).Infof("skipping template %q", template.Name)
+					continue
+				}
+
+				logging.V(10).Infof("adding template %q", template.Name)
+				s.addTemplate(Template{
+					name:               template.Name,
+					description:        "", // No description present.
+					projectDescription: template.Description,
+					source:             s,
+					download: func(ctx context.Context) (workspace.Template, error) {
+						templateDir, err := os.MkdirTemp("", "pulumi-template-")
+						if err != nil {
+							return workspace.Template{}, err
+						}
+						// Having created a template directory, we now add it to the list of directories to close.
+						s.addCloser(func() error { return os.RemoveAll(templateDir) })
+
+						tarReader, err := b.DownloadTemplate(ctx, org, template.TemplateURL)
+						if err != nil {
+							return workspace.Template{}, err
+						}
+						if err := errors.Join(
+							writeTar(ctx, tarReader.Tar(), templateDir),
+							tarReader.Close(),
+						); err != nil {
+							return workspace.Template{}, err
+						}
+						logging.Infof("downloaded %q into %q", template.Name, templateDir)
+
+						return workspace.LoadTemplate(templateDir)
+					},
+				})
+			}
+		}
+	}
+
+	for _, org := range orgs {
+		wg.Add(1)
+		go handleOrg(org)
+	}
+}
+
+func writeTar(_ context.Context, reader *tar.Reader, dst string) error {
```

Comment:
Should we check `context` at each step of the for loop here?  It's pretty cheap to do and `tar` files can have quite horrible characteristics, so being able to abort an unpack might be useful here?

Reply:
Excelent idea.